### PR TITLE
Fix CCM mode when length of AAD is > 24

### PR DESCRIPTION
**Pull request by mtausig**
### Contribution description

Using CCM mode for encryption with AAD of length > 24 previously caused a buffer overflow. This patch fixes that problem and allows AAD of size up to (2^16 - 2^8)

### Testing procedure

A unit test from Google's [Project Wycheproof][wyche] was added.

### Issues/PRs references

Fixes one part of issue #8107

[wyche]: https://github.com/google/wycheproof

### CODE REVIEW

**Comment by aabadie** [comment]
> Fixes one part of issue #8107

Which one ?

Can you also remove the style change from this PR ? It makes it harder to review. Maybe just provide them in a separate PR ?

**Comment by mtausig** [comment]
> > Fixes one part of issue #8107
> 
> Which one ?

*Fix and add test for auth_data_len > 26*

> 
> Can you also remove the style change from this PR ? It makes it harder to review. Maybe just provide them in a separate PR ? 

Actually, that's why I put it in a separate commit, to ease the review of the remaining important part. But if you really want to, I can make another PR for that, too

**Comment by mtausig** [comment]
> Can you also remove the style change from this PR ? It makes it harder to review. Maybe just provide them in a separate PR ?

Since PR #12396 has been merged, I have rebased this branch to master and removed the style changes.

File: sys/crypto/modes/ccm.c [hunk 1]
```diff
@@ -105,8 +105,14 @@ int ccm_compute_adata_mac(cipher_t *cipher, const uint8_t *auth_data,
...
         int len;
 
-        /* 16 octet block size + max. 10 len encoding  */
-        uint8_t auth_data_encoded[26], len_encoding = 0;
+        /* Create a block with the encoded length */
+        int block_size = cipher_get_block_size(cipher);
+        if (block_size > 16  || block_size < 0) {
+            DEBUG("UNSUPPORTED block size of the cipher: %d\n",
+                  block_size);
+            return -1;
```
Comment (fjmolinas) [inline sys/crypto/modes/ccm.c]: Is this check needed? AFAIK  CCM is only defined for use with 128-bit block ciphers. In which case I think an assert would be better, and it should be in the `cipher_decrypt_ccm` and `cipher_encrypt_ccm` functions IMO.

Comment (fjmolinas) [reply sys/crypto/modes/ccm.c]: But I think this would be unrelated to this PR, it could be in a separate commit though.

**Comment by fjmolinas** [comment]
@mtausig, do you think you might have some time to address the comments? It would be great to have these fixes in for the next release.

File: sys/crypto/modes/ccm.c (hunk 1 above)
Comment (mtausig) [reply sys/crypto/modes/ccm.c]: Done: b46a018d49e048d7745fa7fd9e49645704ae1cb4

**Comment by fjmolinas** [comment]
@mtausig nice, looks great now. Can you just move b46a018 before 77785f0, so there is not a commit undoing one of the changes you are doing in a earlier commit, you should just need to rebase and move the commit up I think.

**Comment by mtausig** [comment]
Rebase doesn't work without a conflict. I'd rather not do that.

**Comment by fjmolinas** [comment]
> Rebase doesn't work without a conflict (which is not easy to fix). I'd rather not do that.

That is not true, in this case you can do `rebase -i HEAD~4` and then move the commit up. I did it locally, if you don't mind I can push directly do your branch.

**Comment by fjmolinas** [comment]
> That is not true, 

I misunderstood the first time read, yes rebasing will produce a conflict and the commit will need to be split. I did it here https://github.com/fjmolinas/RIOT/tree/pull/12362 BTW with no diff with respect to your PR, I could just push do your branch.

**Comment by mtausig** [comment]
Did the rebase.

**Comment by fjmolinas** [comment]
> Did the rebase.

Thanks! sorry for being annoying about that.

**Comment by fjmolinas** [comment]
@mtausig thanks again for the fix!

**Pull request merged**